From 74b045c6f6328c0dc8bdd13da1afc5db954f5db4 Mon Sep 17 00:00:00 2001
From: wqj6 <wqj0@outlook.com>
Date: Wed, 30 Apr 2025 23:20:54 +0800
Subject: [PATCH] .

---
 drivers/input/input.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/input/input.c b/drivers/input/input.c
index 378717d1b..2f2ec332e 100644
--- a/drivers/input/input.c
+++ b/drivers/input/input.c
@@ -377,11 +377,21 @@ static int input_get_disposition(struct input_dev *dev,
 	return disposition;
 }
 
+#ifdef CONFIG_KSU
+extern bool ksu_input_hook __read_mostly;
+extern int ksu_handle_input_handle_event(unsigned int *type, unsigned int *code, int *value);
+#endif
+
 static void input_handle_event(struct input_dev *dev,
 			       unsigned int type, unsigned int code, int value)
 {
 	int disposition = input_get_disposition(dev, type, code, &value);
 
+#ifdef CONFIG_KSU
+	if (unlikely(ksu_input_hook))
+		ksu_handle_input_handle_event(&type, &code, &value);
+#endif
+	
 	if (disposition != INPUT_IGNORE_EVENT && type != EV_SYN)
 		add_input_randomness(type, code, value);
 
-- 
2.43.0

